### PR TITLE
create image without pre-installed tf plugins

### LINKED ISSUE #74: Enable dynamic loading of terraform plugins
**How to categorize this issue?**

/area operations
/kind enhancement
/priority normal

**What would you like to be added**:
Add the option to have terraform load necessary provider plugins during runtime. In addition, also generate one image without any pre-installed terraform plugins.

**Why is this needed**:
Currently, there are multiple images being built for the terraformer, containing different sets of pre-installed provider plugins which are then given to the `terraform init` command here: https://github.com/gardener/terraformer/blob/master/pkg/terraformer/terraformer.go#L186
While this is very nice for the Gardener, where there is a fixed scope that allows defining a fixed set of needed terraform plugins, we would like to use the terraformer also for the [landscaper](https://github.com/gardener/landscaper) terraform deployer. This terraform deployer is meant to be able to execute arbitrary terraform operations, which means that it cannot be limited to a pre-defined set of necessary plugins. 
As the `-plugin-dir` argument mentioned above disables the dynamic installation of missing terraform plugins, we would like to have the option - e.g. in form of an command-line argument to the terraformer - to remove it, causing terraform to install the missing providers on demand. 
It would also be nice to have a terraformer image without any pre-installed plugins available, as we don't need them anyway for this scenario.

### PR DESCRIPTION
**Pull request by Diaphteiros**
**How to categorize this PR?**

/kind enhancement
/priority normal

**What this PR does / why we need it**:
As discussed in https://github.com/gardener/terraformer/issues/74, we would like a terraformer image without any pre-installed terraform plugins.

**Which issue(s) this PR fixes**:
Fixes #74 

**Special notes for your reviewer**:
Not sure whether `terraformer-none` or `terraformer-empty` is the more intuitive name here. WDYT?

**Release note**:

```other operator
Terraformer will now publish an additional image without any pre-installed terraform plugins.
```

### CODE REVIEW

**Comment by Diaphteiros** [comment]
cc @schrodit @timebertt

File: build/none/TF_VERSION [hunk 1]
```diff
@@ -0,0 +1 @@
+0.12.29
```
Comment (rfranzke) [inline build/none/TF_VERSION:1]: This should be a symlink

File: build/none/terraform-bundle.hcl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
```
Comment (rfranzke) [inline build/none/terraform-bundle.hcl:1]:
```suggestion
# SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
```

**Review by rfranzke — COMMENTED** [review]
I think you have to extend https://github.com/gardener/terraformer/blob/master/.ci/pipeline_definitions#L10-L50

File: .ci/pipeline_definitions [hunk 3]
```diff
@@ -48,6 +48,12 @@ terraformer:
             target: terraformer
             build_args:
               PROVIDER: equinixmetal
+          terraformer-none:
+            registry: 'gcr-readwrite'
+            image: 'eu.gcr.io/gardener-project/gardener/terraformer-none'
```
Comment (timebertt) [inline .ci/pipeline_definitions:53]: Don't know how to feel about `none`...
WDYT about `slim` or `standalone` or something similar indicating that there are no additional things installed in the image?
Naming is always hard. Don't have a clear preference here.

**Review by timebertt — COMMENTED** [review]
Unsure about the naming, rest lgtm

File: .ci/pipeline_definitions (hunk 3 above)
Comment (Diaphteiros) [reply .ci/pipeline_definitions:53]: I like `slim`, will change it to that unless someone objects.

**Comment by Diaphteiros** [comment]
@timebertt I renamed the image to `terraformer-slim`, but I kept the `none` in the Makefile - is this ok or should I rename it there too?
Should I squash this PR into one commit or will you do it during merging?

**Comment by timebertt** [comment]
Yes, please rename it everywhere consistently to avoid confusion.

> Should I squash this PR into one commit or will you do it during merging?

I've already added the `merge/squash` label which indicates that the person merging the PR should use the "Squash and Merge" option.

**Comment by Diaphteiros** [comment]
> Yes, please rename it everywhere consistently to avoid confusion.

Done.

File: .ci/pipeline_definitions [hunk 4]
```diff
@@ -48,6 +48,12 @@ terraformer:
             target: terraformer
             build_args:
               PROVIDER: equinixmetal
+          terraformer-slim:
+            registry: 'gcr-readwrite'
+            image: 'eu.gcr.io/gardener-project/gardener/terraformer-slim'
+            target: terraformer
+            build_args:
+              PROVIDER: none
```
Comment (timebertt) [inline .ci/pipeline_definitions:56]:
```suggestion
              PROVIDER: slim
```

**Review by timebertt — CHANGES_REQUESTED** [review]

File: .ci/pipeline_definitions (hunk 4 above)
Comment (Diaphteiros) [reply .ci/pipeline_definitions]: Damn, how did I not see that one? Fixed it.

**Review by timebertt — APPROVED** [review]
/lgtm

**Comment by timebertt** [comment]
Image build fails: https://concourse.ci.gardener.cloud/teams/gardener/pipelines/terraformer-master/jobs/master-head-update-job/builds/23#L603545bb:206:209
```text
Step 22/24 : COPY --from=terraform-base /tmp/terraformer/terraform /bin/terraform
 ---> f40151d11bff
Step 23/24 : COPY --from=terraform-base /tmp/terraformer/terraform-provider* /terraform-providers/
COPY failed: no source files were specified
```

@Diaphteiros can you please fix it and check that it works locally using `docker build` (buildkit seems to not fail builds if `COPY` has no source files)?

**Comment by Diaphteiros** [comment]
@timebertt I just tried it locally, but `make docker-images` as well as `docker build` both run through without any problems on my system. I will have a look at the pipeline.

**Comment by Diaphteiros** [comment]
Ah, it's the publish step that fails. Will check that.

**Pull request merged**